tools/please_pex/pex/pex_main.py

```diff
@@ -175,18 +168,22 @@ def __init__(self, module_dir=MODULE_DIR):
 
     def _find_all_distributions(self, module_dir):
         dists = {}
-        if zipfile.is_zipfile(sys.argv[0]):
-            zf = ZipFileWithPermissions(sys.argv[0])
+        pex_file = sys.argv[0]
+        if zipfile.is_zipfile(pex_file):
+            zf = ZipFileWithPermissions(pex_file)
+            r = re.compile(r"{module_dir}{sep}([^/]+)-[^/-]+?\.(?:dist|egg)-info/(.*)".format(
+                module_dir=module_dir,
+                sep = os.sep,
+            ))
             for name in zf.namelist():
-                if name and (m := re.search(
-                    r"{module_dir}(?P<name>[^/]+?)-[^/-]+?\.(?:dist|egg)-info/$".format(
-                        module_dir=module_dir + os.sep,
-                    ),
-                    name,
-                )):
-                    dists.setdefault(m.group("name"), []).append(
-                        PexDistribution(m.group("name"), sys.argv[0], prefix=module_dir)
-                    )
+                match = r.match(name)
+                if match:
+                    mod = match.group(1)
+                    dist = dists.get(mod)
+                    if not dist:
+                        dist = [PexDistribution(mod, pex_file, zf, prefix=module_dir)]
+                        dists[mod] = dist
+                    dist[0]._files[match.group(2)] = name
```

Comment:
I'm not a fan of directly manipulating a private member of another class. Would it be suboptimal to build the entire set of file mappings before constructing the `PexDistribution`s, or to make the `PexDistribution` construct the file mapping itself?

Reply:
Yeah fair. It is a bit cheeky, although it's also a 'local' class so arguably they are allowed to know about each other a bit.
Anyway, have built up the files then gone back to do the distributions in one pass; I think it is a bit nicer like this.